Wrappers/Python/cil/optimisation/algorithms/FISTA.py

```diff
@@ -1,127 +1,273 @@
+            if isinstance(self.f.L, Number):
+                self._step_size = 0.99*2.0/self.f.L
             else:
-                raise ValueError('{} received both initial and the deprecated x_init parameter. It is not clear which one we should use.'\
-                    .format(self.__class__.__name__))
+                raise ValueError("Function f is not differentiable")
+        else:
+            self._step_size = step_size            
         
-        if initial is not None and f is not None:
-            self.set_up(initial=initial, f=f, g=g)
+    def __init__(self, initial, f, g, step_size = None, **kwargs):
 
-    def set_up(self, initial, f, g=ZeroFunction()):
-        '''initialisation of the algorithm
+        super(ISTA, self).__init__(**kwargs)
+        self._step_size = None
+        self.set_up(initial=initial, f=f, g=g, step_size=step_size, **kwargs)
 
-        :param initial: Initial guess ( Default initial = 0)
-        :param f: Differentiable function
-        :param g: Convex function with " simple " proximal operator'''
+    def set_up(self, initial, f, g, step_size, **kwargs):
+        """ Set up of the algorithm
+        """
 
-        logging.info("{} setting up".format(self.__class__.__name__, ))
-        
-        self.y = initial.copy()
+        # set up ISTA      
+        self.initial = initial
         self.x_old = initial.copy()
-        self.x = initial.copy()
-        self.u = initial.copy()
-
+        self.x = initial.copy()           
         self.f = f
         self.g = g
-        if f.L is None:
-            raise ValueError('Error: Fidelity Function\'s Lipschitz constant is set to None')
-        self.invL = 1/f.L
-        self.t = 1
-        self.configured = True
+
+        # set step_size
+        self.set_step_size(step_size=step_size)
+
+        logging.info("{} setting up".format(self.__class__.__name__, ))
+        
+        self.configured = True  
+
         logging.info("{} configured".format(self.__class__.__name__, ))
+              
 
-            
     def update(self):
-        self.t_old = self.t
-        self.f.gradient(self.y, out=self.u)
 
-        self.u.sapyb(-self.invL, self.y, 1.0, out=self.u)
+        r"""Performs a single iteration of ISTA
 
-        self.g.proximal(self.u, self.invL, out=self.x)
-        
-        self.t = 0.5*(1 + numpy.sqrt(1 + 4*(self.t_old**2)))
-        
-        self.x.subtract(self.x_old, out=self.y)
-        self.y.sapyb(((self.t_old-1)/self.t), self.x, 1, out=self.y)
+        .. math:: x_{k+1} = \mathrm{prox}_{\alpha g}(x_{k} - \alpha\nabla f(x_{k}))
 
-        # update_previous_solution() called after update by base class
-        #i.e current solution is now in x_old, previous solution is now in x       
+        """
 
-    def update_previous_solution(self):
-        # swap the pointers to current and previous solution
+        # gradient step
+        self.x_old.sapyb(1., self.f.gradient(self.x_old), -self.step_size, out=self.x)
```

Comment:
Here 3 use more memory than is necessary, we havd `x` `x_old` and the output of `self.f.gradient(self.x_old)`

```python
# gradient step
self.x_old.sapyb(1., self.f.gradient(self.x_old), -self.step_size, out=self.x)
# proximal step
self.x = self.g.proximal(self.x, self.step_size)
```

could be
```python
# gradient step

self.f.gradient(self.x_old, out=self.x)
self.x_old.sapyb(1., self.x, -self.step_size, out=self.x_old)
# proximal step
self.x = self.g.proximal(self.x_old, self.step_size)
```

Reply:
Fixed